app/models/manageiq/providers/ibm_cloud/power_virtual_servers/cloud_manager/event_parser.rb

```diff
@@ -0,0 +1,32 @@
+module ManageIQ::Providers::IbmCloud::PowerVirtualServers::CloudManager::EventParser
+  def self.event_to_hash(event, ems_id)
+    event_hash = {
+      :event_type => "#{event[:resource]}.#{event[:action]}",
+      :source     => "IBMCloud-PowerVS",
+      :ems_id     => ems_id,
+      :ems_ref    => event[:eventID],
+      :timestamp  => event[:time],
+      :full_data  => event
+    }
+
+    _log.debug("event_hash=#{event_hash}")
+    case event_hash[:event_type]
+    when /^pvm-instance/
+      parse_vm_event!(event, event_hash)
+    end
+
+    event_hash
+  end
+
+  def self.parse_vm_event!(event, event_hash)
+    pvm_instance_name = event[:message].split('\'')[1]
+    vm = Vm.find_by(:name => pvm_instance_name)
```

Comment:
VM name is not unique globally, should be `ext_management_system.{vms/vms_and_templates}.find_by(:name => pvm_instance_name)`

You can do the `ems = ExtManagementSystem.find(ems_id)` up in `self.event_to_hash` and pass it down

Reply:
Do you mean include it in the `event_hash` hash? I tried adding an `:ems` key with the value set to the appropriate ems object, but got the following error:
```
[----] I, [2021-11-11T13:50:48.932606 #3797567:f33964]  INFO -- evm: MIQ(MiqQueue#deliver) Message id: [1626], Delivering...                                                                                                                                  
[----] E, [2021-11-11T13:50:50.845219 #3797567:f33964] ERROR -- evm: MIQ(MiqQueue#deliver) Message id: [1626], Error: [unknown attribute 'ems' for EmsEvent.]                                                                                                 
[----] E, [2021-11-11T13:50:50.845390 #3797567:f33964] ERROR -- evm: [ActiveModel::UnknownAttributeError]: unknown attribute 'ems' for EmsEvent.  Method:[block (2 levels) in <class:LogProxy>]                                                               
[----] E, [2021-11-11T13:50:50.845554 #3797567:f33964] ERROR -- evm: /home/jwcarman/.rbenv/versions/2.6.6/lib/ruby/gems/2.6.0/gems/activemodel-6.0.4.1/lib/active_model/attribute_assignment.rb:52:in `_assign_attribute'
```

Instead I grabbed the `:ems_id` from the `event_hash` and add the lookup to `self.parse_vm_event`. This works, but will have to run the query every time in parses a vm event.

Reply:
No sorry, suggesting something like this:
```
def self.event_to_hash(event, ems_id)
  ems = ExtManagementSystem.find(ems_id)
  ...
      parse_vm_event!(ems, event, event_hash)
end

def self.parse_vm_event!(ems, event, event_hash)
  ems.vms.find_by(:name => pvm_instance_name)
```

What you have https://github.com/ManageIQ/manageiq-providers-ibm_cloud/pull/279/files#diff-94f8045f8f587fbabedbe1ade93e3282f34b17895fd96a490f4a095aa5d697edR24 works too but if we end up having more `parse_{resource}_event` methods we might have to do that query in each one